compiler-plugin/src/main/java/io/ballerina/lib/data/yaml/compiler/YamlDataTypeValidator.java

```diff
@@ -0,0 +1,331 @@
+    }
+
+    private void validateExpectedType(TypeSymbol typeSymbol, SyntaxNodeAnalysisContext ctx) {
+        typeSymbol.getLocation().ifPresent(location -> currentLocation = location);
+        switch (typeSymbol.typeKind()) {
+            case UNION -> validateUnionType((UnionTypeSymbol) typeSymbol, ctx);
+            case RECORD -> validateRecordType((RecordTypeSymbol) typeSymbol, ctx);
+            case ARRAY -> validateExpectedType(((ArrayTypeSymbol) typeSymbol).memberTypeDescriptor(), ctx);
+            case TUPLE -> validateTupleType((TupleTypeSymbol) typeSymbol, ctx);
+            case TABLE, XML -> reportDiagnosticInfo(ctx, typeSymbol.getLocation(),
+                    YamlDataDiagnosticCodes.UNSUPPORTED_TYPE);
+            case TYPE_REFERENCE -> validateExpectedType(((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor(), ctx);
+            case INTERSECTION -> validateExpectedType(getRawType(typeSymbol), ctx);
+            default -> { }
+        }
+    }
+
+    private void validateTupleType(TupleTypeSymbol tupleTypeSymbol, SyntaxNodeAnalysisContext ctx) {
+        for (TypeSymbol memberType : tupleTypeSymbol.memberTypeDescriptors()) {
+            validateExpectedType(memberType, ctx);
+        }
+    }
+
+    private void validateRecordType(RecordTypeSymbol recordTypeSymbol, SyntaxNodeAnalysisContext ctx) {
+        detectDuplicateFields(recordTypeSymbol, ctx);
+
+        for (Map.Entry<String, RecordFieldSymbol> entry : recordTypeSymbol.fieldDescriptors().entrySet()) {
+            RecordFieldSymbol fieldSymbol = entry.getValue();
+            currentLocation = fieldSymbol.getLocation().orElseGet(() -> currentLocation);
+            validateExpectedType(fieldSymbol.typeDescriptor(), ctx);
+        }
+    }
+
+    private void validateUnionType(UnionTypeSymbol unionTypeSymbol,
+                                   SyntaxNodeAnalysisContext ctx) {
+        List<TypeSymbol> memberTypeSymbols = unionTypeSymbol.memberTypeDescriptors();
+        for (TypeSymbol memberTypeSymbol : memberTypeSymbols) {
+            validateExpectedType(getRawType(memberTypeSymbol), ctx);
+        }
+    }
+
+    public static TypeSymbol getRawType(TypeSymbol typeDescriptor) {
+        if (typeDescriptor.typeKind() == TypeDescKind.INTERSECTION) {
+            return getRawType(((IntersectionTypeSymbol) typeDescriptor).effectiveTypeDescriptor());
+        }
+        if (typeDescriptor.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+            TypeReferenceTypeSymbol typeRef = (TypeReferenceTypeSymbol) typeDescriptor;
+            if (typeRef.typeDescriptor().typeKind() == TypeDescKind.INTERSECTION) {
+                return getRawType(((IntersectionTypeSymbol) typeRef.typeDescriptor()).effectiveTypeDescriptor());
+            }
+            TypeSymbol rawType = typeRef.typeDescriptor();
+            if (rawType.typeKind() == TypeDescKind.TYPE_REFERENCE) {
+                return getRawType(rawType);
+            }
```

Comment:
We can just call ` getRawType(typeRef.typeDescriptor());` instead of this if condition within L235-L241